src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,164 @@
+import classNames from 'classnames';
+import React, {
+  FormEvent, useEffect, useRef, useState,
+} from 'react';
+import { removeTodo, updateTodo } from '../../api/todos';
+import { Error, Todo, TodoForServer } from '../../types/todo';
+
+type Props = {
+  todo: Todo;
+  todos?: Todo[];
+  setTodos?: (todos: Todo[]) => void;
+  setHasError?: (value: Error) => void;
+  isProcessing?: boolean;
+};
+
+export const TodoItem: React.FC<Props> = ({
+  todo,
+  setTodos = () => {},
+  todos = [],
+  setHasError = () => {},
+  isProcessing = false,
+}) => {
+  const [isHovered, setIsHovered] = useState(false);
+  const [isEdited, setIsEdited] = useState(false);
+  const [editedTitle, setEditedTitle] = useState(todo.title);
+  const [isDeleting, setIsDeleting] = useState(false);
+  const [isToggling, setIsToggling] = useState(false);
+
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (inputRef.current) {
+      inputRef.current.focus();
+    }
+  }, [isEdited]);
+
+  const handleClickEsc = (e: React.KeyboardEvent) => {
+    if (e.key === 'Escape') {
+      setEditedTitle(todo.title);
+      setIsEdited(false);
+    }
+  };
+
+  const updateTodoHandler = (todoId: number, data: TodoForServer) => {
+    updateTodo(todoId, data)
+      .then(() => {
+        const newTodos = todos.map(t => {
+          if (t.id === todoId) {
+            return {
+              ...t,
+              ...data,
+            };
+          }
+
+          return t;
+        });
+
+        setTodos(newTodos);
+      })
+      .catch(() => {
+        setHasError(Error.Update);
+      })
+      .finally(() => {
+        setIsToggling(false);
+        setIsEdited(false);
+      });
+  };
+
+  const removeTodoHandler = (todoId: number) => {
+    setIsDeleting(true);
+
+    removeTodo(todoId)
+      .then(() => {
+        const newTodos = todos.filter(t => t.id !== todoId);
+
+        setTodos(newTodos);
```

Comment:
better to use callback 
```suggestion
        setTodos((prev) => prev.filter(t => t.id !== todoId));
```